hpy/devel/src/runtime/ctx_type.c

```diff
@@ -409,8 +409,17 @@ ctx_Type_FromSpec(HPyContext ctx, HPyType_Spec *hpyspec,
     else {
         // HPyPure_PyObject_HEAD_SIZE ensures that the custom struct is
         // correctly aligned.
-        basicsize = hpyspec->basicsize + HPyPure_PyObject_HEAD_SIZE;
-        base_member_offset = HPyPure_PyObject_HEAD_SIZE;
+        if (hpyspec->basicsize != 0) {
+            basicsize = hpyspec->basicsize + HPyPure_PyObject_HEAD_SIZE;
+            base_member_offset = HPyPure_PyObject_HEAD_SIZE;
+        }
+        else {
+            // If basicsize is 0, it is inherited from the parent type.
+            // Calling HPy_Cast on inherited type only makes sense if the
+            // parent type is already an HPy extension type.
+            basicsize = 0;
+            base_member_offset = 0;
+        }
```

Comment:
The situation with inherited types is quite complicated and I should update this comment and probably turn it into an issue and a todo.

Some of the cases:

* a legacy type inherits the struct of a legacy type: should work.
* a pure type inherits the struct of a pure type: should work right now (both start with PyObject_HEAD plus padding).
* a legacy type inherits from a pure type: should raise an error -- the memory layout definitely won't be right.
* a pure type inherits from a legacy type: works if the legacy type has only PyObject_HEAD, probably fails currently in all other cases.

How to have a pure type inherit from other builtin types -- e.g. PyType_Long -- is an open question that I'd like to leave for another PR. Currently the memory layout will possibly be wrong because HPyPure_PyObject_HEAD_SIZE might include more padding than exists within PyLongObject.

Reply:
I've written up the current situation in https://github.com/hpyproject/hpy/issues/169.

Reply:
I've now also implemented some of the checks that we can make and added tests for them.

I might be possible to add a feature to debug mode where in `HPy_AsStruct` we check that the object we're calling `HPy_AsStruct` on  correctly supports that (although we'd likely need the ability to introspec the type a lot more via the HPy API, so I'd like to leave that as future work).

Reply:
> I might be possible to add a feature to debug mode where in HPy_AsStruct we check that the object we're calling HPy_AsStruct on correctly supports that (although we'd likely need the ability to introspec the type a lot more via the HPy API, so I'd like to leave that as future work).

I'm fine to leave it as future work but maybe open an issue so that we don't forget?